File: java/src/main/java/dev/tslib/Identity.java
package dev.tslib;

/**
 * A TeamSpeak 3 identity based on ECDH.
 *
 * <p>Identities include a security level that must meet server requirements.
 * This class is not thread-safe.
 *
 * <pre>{@code
 * try (Identity id = new Identity()) {
 *     System.out.println("UID: " + id.getUniqueId());
 *     id.save("identity.ini");
 * }
 * }</pre>
 */
public class Identity implements AutoCloseable {
    private long nativePtr;

    /**
     * Create a new random identity.
     *
     * @throws TsLibException if creation fails
     */
    public Identity() {
        this.nativePtr = nativeCreate();
        if (this.nativePtr == 0) {
            throw new TsLibException("Failed to create identity");
        }
    }

    private Identity(long ptr) {
        this.nativePtr = ptr;
    }

    /**
     * Load an identity from a file.
     *
     * @param path file path
     * @return the loaded identity
     * @throws TsLibException if loading fails
     */
    public static Identity load(String path) {
        long ptr = nativeLoad(path);
        if (ptr == 0) {
            throw new TsLibException("Failed to load identity from " + path);
        }
        return new Identity(ptr);
    }

    /**
     * Save the identity to a file.
     *
     * @param path file path
     * @throws TsLibException on error
     */
    public void save(String path) {
        checkNotClosed();
        nativeSave(nativePtr, path);
    }

    /**
     * Import an identity from a string.
     *
     * @param data serialized identity
     * @return the imported identity
     * @throws TsLibException if the format is invalid
     */
    public static Identity fromString(String data) {
        long ptr = nativeFromString(data);
        if (ptr == 0) {
            throw new TsLibException("Failed to parse identity string");
        }
        return new Identity(ptr);
    }

    /**
     * Export the identity as a string.
     *
     * @return serialized identity
     * @throws TsLibException on error
     */
    public String exportString() {
        checkNotClosed();
        return nativeExportString(nativePtr);
    }

    /**
     * Get the unique identifier (public key hash).
     *
     * @return the UID string
     */
    public String getUniqueId() {
        checkNotClosed();
        return nativeGetUniqueId(nativePtr);
    }

    /**
     * Get the current security level.
     *
     * @return security level (0+)
     */
    public int getSecurityLevel() {
        checkNotClosed();
        return nativeGetSecurityLevel(nativePtr);
    }

    /**
     * Get the nickname associated with this identity.
     *
     * @return the nickname, or {@code null} if not set
     */
    public String getNickname() {
        checkNotClosed();
        return nativeGetNickname(nativePtr);
    }

    /**
     * Set the nickname for this identity.
     *
     * @param name the new nickname
     */
    public void setNickname(String name) {
        checkNotClosed();
        nativeSetNickname(nativePtr, name);
    }

    /**
     * Improve the security level. This is CPU-intensive and may block.
     *
     * @param targetLevel desired security level
     * @throws TsLibException on error
     */
    public void improve(int targetLevel) {
        checkNotClosed();
        nativeImprove(nativePtr, targetLevel);
    }

    /**
     * Return the native pointer for use by {@link Client}.
     */
    long getNativePtr() {
        checkNotClosed();
        return nativePtr;
    }

    @Override
    public void close() {
        if (nativePtr != 0) {
            nativeDestroy(nativePtr);
            nativePtr = 0;
        }
    }

    @Override
    public String toString() {
        if (nativePtr == 0) return "Identity(closed)";
        return "Identity(uid=" + getUniqueId() + ", level=" + getSecurityLevel() + ")";
    }

    private void checkNotClosed() {
        if (nativePtr == 0) {
            throw new IllegalStateException("Identity has been closed");
        }
    }

    // Native methods
    private static native long nativeCreate();
    private static native void nativeDestroy(long ptr);
    private static native long nativeLoad(String path);
    private static native void nativeSave(long ptr, String path);
    private static native long nativeFromString(String data);
    private static native String nativeExportString(long ptr);
    private static native String nativeGetUniqueId(long ptr);
    private static native int nativeGetSecurityLevel(long ptr);
    private static native String nativeGetNickname(long ptr);
    private static native void nativeSetNickname(long ptr, String name);
    private static native void nativeImprove(long ptr, int targetLevel);
}

File: crates/tslib-core/src/identity.rs
//! Identity management for TeamSpeak 3
//!
//! TeamSpeak identities are based on ECDH (Elliptic Curve Diffie-Hellman)
//! and include a security level that must meet server requirements.

use crate::error::{IdentityError, Result};
use serde::{Deserialize, Serialize};
use std::path::Path;

// Re-export tsclientlib's identity for internal use
pub use tsclientlib::Identity as TsIdentity;

/// A TeamSpeak 3 identity wrapper
///
/// Wraps tsclientlib's Identity with additional convenience methods.
#[derive(Clone)]
pub struct Identity {
    /// The underlying tsclientlib identity
    inner: TsIdentity,
    /// Optional nickname associated with this identity
    nickname: Option<String>,
}

impl Identity {
    /// Create a new random identity
    ///
    /// The identity starts at security level 0 and should be improved
    /// before connecting to servers with security requirements.
    pub fn create() -> Result<Self> {
        let inner = TsIdentity::create();
        Ok(Self {
            inner,
            nickname: None,
        })
    }

    /// Create an identity from the underlying tsclientlib identity
    pub fn from_ts_identity(inner: TsIdentity) -> Self {
        Self {
            inner,
            nickname: None,
        }
    }

    /// Get the underlying tsclientlib identity
    pub fn as_ts_identity(&self) -> &TsIdentity {
        &self.inner
    }

    /// Get a clone of the underlying tsclientlib identity
    pub fn to_ts_identity(&self) -> TsIdentity {
        self.inner.clone()
    }

    /// Load an identity from a file
    pub fn load(path: impl AsRef<Path>) -> Result<Self> {
        let content = std::fs::read_to_string(path.as_ref())
            .map_err(|e| IdentityError::ImportFailed(e.to_string()))?;

        Self::from_string(&content)
    }

    /// Save the identity to a file
    pub fn save(&self, path: impl AsRef<Path>) -> Result<()> {
        let content = self.export_string()?;
        std::fs::write(path.as_ref(), content)
            .map_err(|e| IdentityError::ExportFailed(e.to_string()))?;
        Ok(())
    }

    /// Import an identity from a base64-encoded string
    pub fn from_string(data: &str) -> Result<Self> {
        // Try to parse as our JSON format first
        if let Ok(stored) = serde_json::from_str::<StoredIdentity>(data) {
            let inner = TsIdentity::new_from_str(&stored.key)
                .map_err(|e| IdentityError::ImportFailed(e.to_string()))?;
            return Ok(Self {
                inner,
                nickname: stored.nickname,
            });
        }

        // Try to parse as raw tsclientlib format
        let inner = TsIdentity::new_from_str(data.trim())
            .map_err(|e| IdentityError::ImportFailed(e.to_string()))?;

        Ok(Self {
            inner,
            nickname: None,
        })
    }

    /// Export the identity to a string
    pub fn export_string(&self) -> Result<String> {
        // Export the key in tsclientlib format
        let key = self.export_key_string();
        let stored = StoredIdentity {
            key,
            nickname: self.nickname.clone(),
        };
        serde_json::to_string_pretty(&stored)
            .map_err(|e| IdentityError::ExportFailed(e.to_string()).into())
    }

    /// Export just the key string in tsclientlib format
    fn export_key_string(&self) -> String {
        // Format: counterVbase64(private_key_bytes)
        use base64::Engine;
        let key_bytes = self.inner.key().to_short();
        let counter = self.inner.counter();
        let key_b64 = base64::engine::general_purpose::STANDARD.encode(&key_bytes);
        format!("{}V{}", counter, key_b64)
    }

    /// Get the unique identifier (public key hash)
    pub fn unique_id(&self) -> String {
        self.inner.key().to_pub().get_uid()
    }

    /// Get the current security level
    pub fn security_level(&self) -> u8 {
        self.inner.level()
    }

    /// Get the key counter (offset)
    pub fn key_offset(&self) -> u64 {
        self.inner.counter()
    }

    /// Get the associated nickname, if any
    pub fn nickname(&self) -> Option<&str> {
        self.nickname.as_deref()
    }

    /// Set the nickname for this identity
    pub fn set_nickname(&mut self, nickname: impl Into<String>) {
        self.nickname = Some(nickname.into());
    }

    /// Improve the identity's security level synchronously
    ///
    /// This is a CPU-intensive operation that finds a key offset
    /// which produces a hash with enough leading zeros.
    ///
    /// # Arguments
    /// * `target_level` - The desired security level (8-32 typically)
    pub fn improve(&mut self, target_level: u8) -> Result<()> {
        if target_level <= self.inner.level() {
            return Ok(());
        }

        self.inner.upgrade_level(target_level);
        Ok(())
    }

    /// Improve the identity asynchronously
    ///
    /// Runs the CPU-intensive work in a blocking task pool.
    pub async fn improve_async(&mut self, target_level: u8) -> Result<()> {
        if target_level <= self.inner.level() {
            return Ok(());
        }

        let mut inner = self.inner.clone();

        let result = tokio::task::spawn_blocking(move || {
            inner.upgrade_level(target_level);
            inner
        })
        .await
        .map_err(|e| IdentityError::ImproveFailed(e.to_string()))?;

        self.inner = result;
        Ok(())
    }
}

impl std::fmt::Debug for Identity {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Identity")
            .field("unique_id", &self.unique_id())
            .field("security_level", &self.security_level())
            .field("nickname", &self.nickname)
            .finish_non_exhaustive()
    }
}

/// Stored identity format for serialization
#[derive(Debug, Serialize, Deserialize)]
struct StoredIdentity {
    /// The identity key in tsclientlib format
    key: String,
    /// Optional nickname
    #[serde(skip_serializing_if = "Option::is_none")]
    nickname: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_create_identity() {
        let identity = Identity::create().unwrap();
        assert!(!identity.unique_id().is_empty());
        assert_eq!(identity.security_level(), 0);
    }

    #[test]
    fn test_serialize_identity() {
        let identity = Identity::create().unwrap();
        let json = identity.export_string().unwrap();
        let loaded = Identity::from_string(&json).unwrap();
        assert_eq!(identity.unique_id(), loaded.unique_id());
    }
}
